Convert Temperature Component to support multiple instances

bradjc opened the pull request:
### Pull Request Overview

Part of #3222 

I thought the temperature component is somewhat simple, so this PR converts the component to (hopefully) support being used multiple times by the same board. I'm hoping we can discuss this PR to get a template for converting all the rest.

The main change here is adding:

```rust
#[macro_export]
macro_rules! temperature_component_static_init {
    () => {{
        kernel::static_buf!(capsules::temperature::TemperatureSensor<'static>)
    };};
}
```
to the component.

I chose the name `temperature_component_static_init` as the purpose of adding this macro is to encapsulate the static_init!() portion which is what needs to be instantiated in the main.rs file itself, not in the finalize() function. The name is not `_helper`, as I created the "_helper" name because those macros helped me create some way to enable component sharing across microcontrollers. I didn't have a better name at the time.

I'm thinking we should revisit the naming convention for the macro. Perhaps we just use `X_static!()` for these macros, to signify this is where the static buf is defined?




### Testing Strategy

travis


### TODO or Help Wanted

Can we do better than this?


### Documentation Updated

- [ ] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [ ] Ran `make prepush`.

hudson-ayers reviewed (dismissed):
I think this PR looks good -- it lets the component be used for multiple instances in the "tock-idiomatic" way. 

I do think that given our experience with components so far we might want to consider making components just be macros in the future. `static_init!()` requires us to use macros for most components in order for them to be reusable anyway, and I have doubts that this half-macro / half-trait approach is any more reader-friendly than a pure-macro approach (which would also be much more brief, both in implementation and instantiation).

bradjc commented:
I changed the name to `temperature_component_static!()`, and I think we should only use that naming convention. The purpose of those macros is to handle the static defines---nothing else.

I also added some documentation to the porting guide.

hudson-ayers commented:
> I changed the name to `temperature_component_static!()`, and I think we should only use that naming convention. The purpose of those macros is to handle the static defines---nothing else.

Agreed this should be the purpose of these macros, though we do not follow this everywhere today (e.g. https://github.com/tock/tock/blob/master/boards/components/src/temperature_stm.rs#L11)

bradjc commented:
> > I changed the name to `temperature_component_static!()`, and I think we should only use that naming convention. The purpose of those macros is to handle the static defines---nothing else.
> 
> Agreed this should be the purpose of these macros, though we do not follow this everywhere today (e.g. https://github.com/tock/tock/blob/master/boards/components/src/temperature_stm.rs#L11)

Yes, working on it #3231

bradjc commented:
I guess I meant to
